### PR TITLE
Reference fields

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR provides a major overhaul for how the "reference" fields are implemented:

- BuildOrder.reference
- SalesOrder.reference
- PurchasesOrder.reference

The *current* implementation provides a "prefix" and "regex" setting for each type of reference, e.g. for a BuildOrder the prefix might be `BO-` and the regex might be `(\d{4})`, meaning that the 'reference field' *must* be something like `BO-1234`

However this is overly complex, and does not allow us to implement the more complex behaviours as explored in https://github.com/inventree/InvenTree/discussions/3163

So this PR simultaneously provides an overhaul, simplification and expansion of functionality.

The "prefix" and "regex" concepts are consolidated into a configurable "reference pattern":

- Uses "pythonic" string formatting for named group replacement
- Allows simple pattern matching
- Allows specification of integer portion of reference

#### Example

Let's say the *Build Order Reference Pattern* is configured to be:

`BO-{ref:04d}`

image

This pattern serves to:

1. Auto-generate the next build reference
2. Enforce a certain pattern for the build reference

Using python format strings allows us to expand this concept in the future, although the pattern substitution functionality is currently very limited.

---------------------------------
- Closes https://github.com/inventree/InvenTree/issues/3167
- Closes https://github.com/inventree/InvenTree/pull/3172

<a href="https://gitpod.io/#https://github.com/inventree/InvenTree/pull/3267"></a>

### TODO

- [x] Data migration for build order reference
- [x] Remove build order prefix setting
- [x] Update the "Sales Order" reference field accordingly
- [x] Update the "Purchase Order" reference field accordingly
- [x] Data migration for "Purchase Order" reference field
- [x] Data migration for "Sales Order" reference field
- [x] Remove purchase order prefix setting
- [x] Remove sales order prefix setting
- [x] Documentation - https://github.com/inventree/inventree-docs/pull/308
- [ ] Update unit tests for `inventree-python`
- [x] Perform migrations on demo dataset - https://github.com/inventree/demo-dataset/pull/26

### Future Extensions

This pattern idea can be also used for

- Serial numbers
- Batch codes
- Shipment codes
- Part name formatting

### CODE REVIEW

**Comment by matmair** [comment]
@SchrodingersGat why not use f-string numerals? https://peps.python.org/pep-0498/ would make the syntax simpler and there are a lot of references. We could provide a context like this:
```
now: current timestamp
customer: current customer
...
serial: reference number
```

The sample string would be formatted like this:
`BO-{now:%Y.%W}-{serial:05d}`

Guide the f-strings https://realpython.com/python-string-formatting/#3-string-interpolation-f-strings-python-36

**Comment by SchrodingersGat** [comment]
Using f-strings would be more powerful, but I feel less "accessible" to the general user, which is why I am moving away from user-facing regular expressions (per the discussion in https://github.com/inventree/InvenTree/pull/3172)

e.g. `PO-#####` vs `PO-{serial:05d}`

**Comment by matmair** [comment]
@SchrodingersGat I'll let you finish before discussing it. Inventing a new pattern-matching system seems like unnecessary work to me.

**Comment by SchrodingersGat** [comment]
@matmair I have updated the description now, but you'll probably want to have a read through the code to see what I'm going for.

Basically I have converted it to use pythonic string formatting, and trying to make a simple mixin class-based approach. LMK what you think. It will certainly need some more refinement.

File: InvenTree/InvenTree/models.py [hunk 1]
```diff
@@ -112,21 +112,170 @@ class ReferenceIndexingMixin(models.Model):
...
+        try:
+            incremented = int(incremented)
+        except ValueError:
+            pass
+
+        return incremented
+
+    @classmethod
+    def generate_reference(cls):
+        """Generate the next 'refernce' field based on specified pattern"""
```
Comment (matmair) [inline InvenTree/InvenTree/models.py:185]: Typo

**Review by matmair — APPROVED** [review]

**Comment by matmair** [comment]
@SchrodingersGat looks good, I added some non-blocking comments but should be good to go from my POV.

File: InvenTree/InvenTree/serializers.py [hunk 2]
```diff
@@ -211,16 +208,6 @@ class Meta:
         ]
 
 
-class ReferenceIndexingSerializerMixin():
-    """This serializer mixin ensures the the reference is not to big / small for the BigIntegerField."""
-
-    def validate_reference(self, value):
-        """Ensures the reference is not to big / small for the BigIntegerField."""
-        if extract_int(value) > models.BigIntegerField.MAX_BIGINT:
```
[the remark this answers is not in the document]
Comment (SchrodingersGat) [reply InvenTree/InvenTree/serializers.py:219]: Has been added back in, thanks for pointing this out

[the remark this answers is not in the document]
Comment (matmair) [reply InvenTree/InvenTree/serializers.py:219]: @SchrodingersGat we had that bug already, no need to refind the solution 😇

**Review by matmair — APPROVED** [review]

**Comment by matmair** [comment]
@SchrodingersGat the settings for orders look pretty sparse - should we just merge them (SO, PO, BO) together?

**Comment by miggland** [comment]
Hey guys, sorry for not reacting until now, just have been busy with other stuff.

This looks really good, a quick test confirms that this is what I was after in my original FR.

**Pull request merged**